### DISCUSSION TITLE
Not being able to install Homebrew

### DISCUSSION THREAD

**Discussion by mmarcelino** [Homebrew/discussions#4309] [Getting started]
### Output of `brew config`

```shell
zsh: command not found: brew
```

### Output of `brew doctor`

```shell
zsh: command not found: brew
```

### Description of issue

I'm trying to install homebrew by inputting `/bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"` on the terminal, but I keep getting the following error message:

```
/bin/bash: line 234: ./bin/git: No such file or directory
Failed during: ./bin/git -c init.defaultBranch=master init --quiet
```
I've googled it, but haven't found any helpful suggestions, so would highly appreciate it if anyone could help me here. I'm on a Mac, using Mojave v. 10.14.6 and I do have git installed (git version 2.20.1 (Apple Git-117)).

Thanks in advance!

**Comment by SMillerDev** [comment]
> I'm on a Mac, using Mojave v. 10.14.6

That's not supported, you'll likely encounter some issues.

> I do have git installed (git version 2.20.1 (Apple Git-117)).

Where is it installed?

**Reply by mmarcelino** [reply]
Thanks for the prompt reply, @SMillerDev. git is under `/usr/bin/git`. Any suggestions? Thank you!

**Reply by SMillerDev** [reply]
What does `echo $PATH` say?

**Reply by mmarcelino** [reply]
It says this: `/Users/marianamarcelino/.rvm/gems/ruby-2.4.3/bin:/Users/marianamarcelino/.rvm/gems/ruby-2.4.3@global/bin:/Users/marianamarcelino/.rvm/rubies/ruby-2.4.3/bin:./bin:./node_modules/.bin:/Users/marianamarcelino/.rbenv/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/Library/Frameworks/Mono.framework/Versions/Current/Commands:/usr/local/sbin:/Users/marianamarcelino/.rvm/bin`

**Comment by gromgit** [comment]
>```
>/bin/bash: line 234: ./bin/git: No such file or directory
>```

You have `./bin` very early in your `PATH`. Relative paths in `PATH` is a **SUPREMELY BAD IDEA**, because what gets run now depends on which directory you're in. HaX0rs *love* this kind of setup, as they can now drop stuff in unexpected places, then LOL as you accidentally type `sl` instead of `ls`...and give them a RAT shell: https://www.howtogeek.com/410634/what-is-rat-malware-and-why-is-it-so-dangerous/.

Since you have a (presumably known-good) Git in `/usr/bin/git`, you can do this to get Homebrew installed:
```
PATH=/usr/bin:$PATH /bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"
```
but you REALLY _REALLY_ _**REALLY**_ should get rid of that `./bin` in your `PATH`. You have been warned.

**Reply by mmarcelino** [reply]
Thank you, @gromgit! 🙏 I have no idea how that happened, as I'm super inexperienced in these matters. Any suggestion re: how I should go about getting rid of that `./bin` in my `PATH`?

**Reply by gromgit** [reply]
Based on the `/Users/...` directories in your `PATH`, I'm assuming you're on a macOS box, so do the following:
```sh
# Run the appropriate shell in login-tracing mode
if [[ -n $ZSH_VERSION ]]; then
  script /tmp/shell.log zsh -xl -c "exit"
elif [[ -n $BASH_VERSION ]]; then
  PS4='+$BASH_SOURCE> ' BASH_XTRACEFD=2 script /tmp/shell.log bash -xl -c "exit"
else
  echo "ERROR: Strange shell encountered ($SHELL)"
fi
# Then see where ./bin is being added
grep -F ./bin /tmp/shell.log
```
That last `grep` will show you the filename and line number of the command that added the `./bin`. It's up to you to edit that file and remove it. (The `./node_modules/bin` I see in your reply to @SMillerDev is also not recommended.)

**Reply by mmarcelino** [reply]
Thank you, @gromgit. When I run  `grep -F ./bin /tmp/shell.log` I get this:
```
+/etc/zprofile:3> eval 'PATH="/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/Library/Frameworks/Mono.framework/Versions/Current/Commands:/Users/marianamarcelino/.rvm/gems/ruby-2.4.3/bin:/Users/marianamarcelino/.rvm/gems/ruby-2.4.3@global/bin:/Users/marianamarcelino/.rvm/rubies/ruby-2.4.3/bin:./bin:./node_modules/.bin:/Users/marianamarcelino/.rbenv/bin:/usr/local/sbin:/Users/marianamarcelino/.rvm/bin";' export 'PATH;'
+(eval):1> PATH=/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/Library/Frameworks/Mono.framework/Versions/Current/Commands:/Users/marianamarcelino/.rvm/gems/ruby-2.4.3/bin:/Users/marianamarcelino/.rvm/gems/ruby-2.4.3@global/bin:/Users/marianamarcelino/.rvm/rubies/ruby-2.4.3/bin:./bin:./node_modules/.bin:/Users/marianamarcelino/.rbenv/bin:/usr/local/sbin:/Users/marianamarcelino/.rvm/bin 
+/Users/marianamarcelino/.rvm/environments/default:1> export PATH=/Users/marianamarcelino/.rvm/gems/ruby-2.4.3/bin:/Users/marianamarcelino/.rvm/gems/ruby-2.4.3@global/bin:/Users/marianamarcelino/.rvm/rubies/ruby-2.4.3/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/Library/Frameworks/Mono.framework/Versions/Current/Commands:/Users/marianamarcelino/.rvm/gems/ruby-2.4.3/bin:/Users/marianamarcelino/.rvm/gems/ruby-2.4.3@global/bin:/Users/marianamarcelino/.rvm/rubies/ruby-2.4.3/bin:./bin:./node_modules/.bin:/Users/marianamarcelino/.rbenv/bin:/usr/local/sbin:/Users/marianamarcelino/.rvm/bin
+__rvm_project_rvmrc:13> save_PATH=/Users/marianamarcelino/.rvm/gems/ruby-2.4.3/bin:/Users/marianamarcelino/.rvm/gems/ruby-2.4.3@global/bin:/Users/marianamarcelino/.rvm/rubies/ruby-2.4.3/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/Library/Frameworks/Mono.framework/Versions/Current/Commands:/Users/marianamarcelino/.rvm/gems/ruby-2.4.3/bin:/Users/marianamarcelino/.rvm/gems/ruby-2.4.3@global/bin:/Users/marianamarcelino/.rvm/rubies/ruby-2.4.3/bin:./bin:./node_modules/.bin:/Users/marianamarcelino/.rbenv/bin:/usr/local/sbin:/Users/marianamarcelino/.rvm/bin 
+/Users/marianamarcelino/.rvm/scripts/rvm:244> [[ :/Users/marianamarcelino/.rvm/gems/ruby-2.4.3/bin:/Users/marianamarcelino/.rvm/gems/ruby-2.4.3@global/bin:/Users/marianamarcelino/.rvm/rubies/ruby-2.4.3/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/Library/Frameworks/Mono.framework/Versions/Current/Commands:/Users/marianamarcelino/.rvm/gems/ruby-2.4.3/bin:/Users/marianamarcelino/.rvm/gems/ruby-2.4.3@global/bin:/Users/marianamarcelino/.rvm/rubies/ruby-2.4.3/bin:./bin:./node_modules/.bin:/Users/marianamarcelino/.rbenv/bin:/usr/local/sbin:/Users/marianamarcelino/.rvm/bin: == *:/Users/marianamarcelino/.rvm/bin:* ]]
```
So, multiple `./bin`. To be honest, and as previously mentioned, I'm pretty inexperienced in these matters, so when you say that it's up to me to edit the file and remove it, I don't exactly know how to go about it.

Thanks in advance!

**Reply by gromgit** [reply]
>```
>+/etc/zprofile:3> eval 'PATH="/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/Library/Frameworks/Mono.framework/Versions/Current/Commands:/Users/marianamarcelino/.rvm/gems/ruby-2.4.3/bin:/Users/marianamarcelino/.rvm/gems/ruby-2.4.3@global/bin:/Users/marianamarcelino/.rvm/rubies/ruby-2.4.3/bin:./bin:./node_modules/.bin:/Users/marianamarcelino/.rbenv/bin:/usr/local/sbin:/Users/marianamarcelino/.rvm/bin";' export 'PATH;'
>```

This is the statement that's adding `./bin` and `./node_modules/.bin` to your `PATH`. What's the output of `sed -n 3p /etc/zprofile`?

**Reply by mmarcelino** [reply]
The output of `sed -n 3p /etc/zprofile` is the following: `eval /usr/libexec/path_helper -s`, @gromgit

**Reply by gromgit** [reply]
You might want to copy-and-paste output in future, instead of manually retyping it. That line should be:
```
eval `/usr/libexec/path_helper -s`
```
not, as you have written:
```
eval /usr/libexec/path_helper -s
```
which does something very different.

In any case, `path_helper` picks up all the paths listed in all the files under `/etc/paths.d`, and pastes them all together into a single `PATH` value. One of those files contains `./bin`, while another one (or perhaps even the same file) contains `./node_modules/.bin`. Those files were put there either by:
* you, following some online tutorial that prioritizes convenience over security, or
* by installing some software or other--I'm guessing you've at least installed Node.js outside of Homebrew.

This is where I stop, because _it's your system_, not mine:
* Any relative path (i.e. any directory that doesn't start with `/`) that was added by a software installer, you want to ask on an support forum for that software whether you can remove it safely.
* Any relative path you added manually, you want to ask yourself _why_ you added it, now you know how that opens a security hole.
